Unpacking RESTful sensor JSON results into attributes.

nickovs opened the pull request:
## Description:

Currently the **RESTful** sensor can only unpack a single JSON value using a template. There are many cases where the REST result contains multiple values. There are also many cases where making the request is costly, for example when accessing commercial or rate-limited APIs, fetching results from tiny, low power IoT devices or making the request across a slow network. As a result it is useful to be able to fetch all of the resulting values in a single request.

This patch adds a `json_attributes` configuration option to the **RESTful** sensor which indicates that the result of the REST request should be treated as a JSON dictionary and the keys in that dictionary should be mapped to attributes in the sensor. This allows all of the resulting values to be accessed in other parts of the system. Furthermore it allows new sensors to be created that monitor the value of individual attributes using the **template** sensor platform.


**Related issue (if applicable):** Addresses the feature request in the community thread ["RESTful Sensor - read multiple values in one request"](https://community.home-assistant.io/t/restful-sensor-read-multiple-values-in-one-request/1654).


**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4046

## Example entry for `configuration.yaml`:

By way of example, the configuration below creates a set of sensors for the temperature, pressure, humidity and weather conditions retrieved from OpenWeatherMap.org using a single REST call.

```
- platform: rest
  name: OWM_report
  json_attributes: True
  value_template: '{{value_json["weather"][0]["description"].title()}}'
  resource: http://api.openweathermap.org/data/2.5/weather?zip=80302,us&APPID=VERYSECRETAPIKEY

- platform: template
  sensors:
    owm_weather:
      value_template: '{{states.sensor.owm_report.status}}'
      icon_template: '{{"http://openweathermap.org/img/w/"+states.sensor.owm_report.attributes.weather[0]["icon"]+".png"}}'
      entity_id: sensor.owm_report
    owm_temp:
      friendly_name: 'Outside temp'
      value_template: '{{states.sensor.owm_report.attributes.main["temp"]-273.15}}'
      unit_of_measurement: "°C"
      entity_id: sensor.owm_report
    owm_pressure:
      friendly_name: 'Outside pressure'
      value_template: '{{states.sensor.owm_report.attributes.main["pressure"]}}'
      unit_of_measurement: "hP"
      entity_id: sensor.owm_test
    owm_humidity:
      friendly_name: 'Outside humidity'
      value_template: '{{states.sensor.owm_report.attributes.main["humidity"]}}'
      unit_of_measurement: "%"
      entity_id: sensor.owm_test
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

imrehg committed:
sensor.envirophat: add missing requirement (#7451)

Adding requirements that is not explicitly pulled in by the library
that manages the Enviro pHAT.

JshWright committed:
PyPI Openzwave (#7415)

* Remove default zwave config path

PYOZW now has much more comprehensive default handling for the config
path (in src-lib/libopenzwave/libopenzwave.pyx:getConfig()). It looks in
the same place we were looking, plus _many_ more. It will certainly do a
much better job of finding the config files than we will (and will be
updated as the library is changed, so we don't end up chasing it). The
getConfig() method has been there for a while, but was subsntially
improved recently.

This change simply leaves the config_path as None if it is not
specified, which will trigger the default handling in PYOZW.

* Install python-openzwave from PyPI

As of version 0.4, python-openzwave supports installation from PyPI,
which means we can use our 'normal' dependency management tooling to
install it. Yay.

This uses the default 'embed' build (which goes and downloads
statically sources to avoid having to compile anything locally). Check
out the python-openzwave readme for more details.

* Add python-openzwave deps to .travis.yml

Python OpenZwave require the libudev headers to build. This adds the
libudev-dev package to Travis runs via the 'apt' addon for Travis.

Thanks to @MartinHjelmare for this fix.

* Update docker build for PyPI openzwave

Now that PYOZW can be install from PyPI, the docker image build process
can be simplified to remove the explicit compilation of PYOZW.

nunofgs committed:
Add datadog component (#7158)

* Add datadog component

* Improve test_invalid_config datadog test

* Use assert_setup_component for test setup

JshWright committed:
Fix object type for default KNX port

#7429 describes a TypeError that is raised if the port is omitted in the config for the KNX component (integer is required (got type str)). This commit changes the default port from a string to an integer. I expect this will resolve that issue...

nickovs committed:
Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

nickovs committed:
Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

nickovs committed:
Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

nickovs committed:
Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

nickovs committed:
Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

nickovs committed:
Added support for extracting JSON attributes from RESTful values

Setting the json_attributes configuration option to true on the
RESTful sensor will cause the result of the REST request to be parsed
as a JSON string and if successful the resulting dictionary will be
used for the attributes of the sensor.

nickovs committed:
Merge branch 'rest-json-attrs' of https://github.com/nickovs/home-assistant into rest-json-attrs

nickovs committed:
Added requirement that RESTful JSON results used as attributes must be
objects, not lists.

nickovs committed:
Added requirement that RESTful JSON results used as attributes must be
objects, not lists.

nickovs committed:
Added requirement that RESTful JSON results used as attributes must be
objects, not lists.

nickovs committed:
Added requirement that RESTful JSON results used as attributes must be
objects, not lists.

nickovs committed:
Added requirement that RESTful JSON results used as attributes must be
objects, not lists.

nickovs committed:
Expanded test coverage to test REFTful JSON attributes with and
without a value template.

nickovs committed:
Expanded test coverage to test REFTful JSON attributes with and
without a value template.

nickovs committed:
Expanded test coverage to test REFTful JSON attributes with and
without a value template.

nickovs committed:
Expanded test coverage to test REFTful JSON attributes with and
without a value template.

nickovs committed:
Expanded test coverage to test REFTful JSON attributes with and
without a value template.

nickovs committed:
Merge branch 'rest-json-attrs' of github.com:nickovs/home-assistant into rest-json-attrs

nickovs committed:
Merge branch 'rest-json-attrs' of github.com:nickovs/home-assistant into rest-json-attrs

nickovs committed:
Brought up to date with most recent upstream master

homeassistant commented:
Hi @nickovs, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#10753).

Once you do that we will be able to review and accept this pull request.

Thanks!

nickovs commented:
The CLA documentation says that to sign the CLA I should submit a pull request and then "a link will be created" and from there I should follow the instructions. I'm seeing a check failure from the cla-bot saying that I have not signed the CLA but I can't find any link. Can someone point me to where the link might be found?

tinloaf commented:
I see a couple of problems with this:
* Most JSON payloads will not just be a dictionary of primitive types. What happens if you have a dictionary in the dictionary? Then you have an attribute that is the string representation of a python dictionary (which may contain other non-primitive types). This will result in all sorts of weird attributes.
* I don't think attributes have the same 255 character limit that the state does (I might be wrong…), but this could result in attributes having kilo- or megabytes of size. This probably wreaks havoc on the recorder or other history components.
* This allows whoever controls the REST service to insert arbitrary strings into the front end of people using a RESTful sensor. I wouldn't be sure that the UI is so battened down that it does HTML escaping everywhere it uses an attribute name.

nickovs committed:
Fixed breaks cause by manual upstream merge.

nickovs commented:
@tinloaf Thanks for the comments. Regarding your concerns:

1. The vast majority of REST API endpoints return JSON dictionaries and this is almost invariably the case for the sort of endpoints that for which people will use this option. If the returned result is not a dictionary then the code will log a warning and ignore the result.
1. It is possible that the REST endpoint will return a large result but the user doesn't have to use this if that's going to be a problem. It would be possible to some arbitrary a cap on the complexity of the results that we might store but I submit that whatever value you choose for a cap like that there will be some people who think it is wrong. This is an _optional_ setting so the user can always choose to do it the old way.
1. The existing RESTful sensor already allows the owner of the REST endpoint to insert content into frontend when people use the returned values. Attributes get escaped during rendering and if you use a template sensor to extract values from the attributes then you're in the same place that the current code puts you.

nickovs committed:
Added one extra blank line to make PyLint happy.

nickovs commented:
@tinloaf: If the size of the JSON result is a serious concern then it would be possible to add another configuration option to specify either some dictionary sub-tree of the JSON object from which to extract values or a set of pairs of attribute names and templates to extract. That said, this seems overly complex to me, both in terms of code complexity and difficulty for the user to configure. The solution here has the advantage of being much simpler on both counts.

tinloaf commented:
I'm not completely sure, someone with more knowledge of the state machine might have a better guess at that than I. The thing is: The state is currently automatically limited to 255 characters. Reading JSON into attributes might result in large amounts of data going through the state machine.

nickovs commented:
@tinloaf: If the concern is about the size of data put into the state value then I could change the code to suppress the JSON result from going into the state value directly (but still allow it through a template). The attributes are Python objects, not strings, so there is no size restriction. Looking at the `events` and `states` recorder database tables the database does not seem to record the full set of attributes, only the state values, so the volume of data that ends up in the database is exactly the same as if you use multiple RESTful sensors hitting the same REST API multiple times.

pvizeli commented on homeassistant/components/sensor/rest.py:
```diff
@@ -68,20 +72,25 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
         self._value_template = value_template
+        self._json_attrs = json_attrs
+        self._attributes = None
+        self.update()
```
Remove this

nickovs replied:
When you say "this", presumably you mean the `self.update()` rather than the whole section.

pvizeli commented on homeassistant/components/sensor/rest.py:
```diff
@@ -68,20 +72,25 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
 
-    add_devices([RestSensor(hass, rest, name, unit, value_template)], True)
+    add_devices([RestSensor(hass, rest, name, unit,
+                            value_template, json_attrs)])
```
Revert the True at the end

nickovs replied:
Sorry about that. The flag was added after I forked the code and I missed that in the merge.

pvizeli reviewed (dismissed):
Make `json_attr` as list to select which attribute do you want

nickovs committed:
Switched json_attributes to be a list of keys rather than a boolean.

The value of json_attributes can now be either a comma sepaated list
of key names or a YAML list of key names. Only matching keys in a
retuned JSON dictionary will be mapped to sensor attributes.

Updated test cases to handle json_attributes being a list.

Also fixed two minor issues arrising from manual merge with 0.58 master.

nickovs commented:
@pvizeli I've changed the code so that the `json_attributes` configuration parameter is now a list of keys to fetch from the a returned JSON dictionary that will be mapped as attributes of the sensor, instead of just being a boolean. I used the `ensure_list_csv()` function from the config validation helper code, so users can provide the list either as a YAML list or as a comma separated string.

The test cases have been updated accordingly and I've also pushed updated documentation.

The new version also fixes the two issues that you picked up that resulted from my manual merge to bring the base up to date with the 0.58 codebase.

Let me know if you have any other suggestions or requests.

fabaff commented on homeassistant/components/sensor/rest.py:
```diff
@@ -40,6 +42,7 @@
     vol.Optional(CONF_USERNAME): cv.string,
     vol.Optional(CONF_VALUE_TEMPLATE): cv.template,
     vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
+    vol.Optional(CONF_JSON_ATTRS): cv.ensure_list_csv,
```
According to your docs the default is an empty list (-> `default=[]`). To be in sync with other platform we should only accept a list and not a comma-separated string ->`vol.Optional(CONF_JSON_ATTRS): vol.All(cv.ensure_list, [cv.string]),`

nickovs replied:
I've added the default (although the existing code treats the a value of `None` the same as an empty list).

Regarding the use of `cv.ensure_list_csv`, is this function deprecated? The function seems to be there explicitly to allow users to enter lists either as a YAML list or as a CVS list, which is very helpful to users and allows them to choose the representation that they find most readable and compact. The function gets used by calendar/todoist.py, insteon_plm, sensor/mvglive.py and switch/rpi_rf.py and serves a very useful purpose. If the intention is to do way with this useful capability then I can remove it from here but if the function is going to stay around is there really a compelling reason not to use it?

fabaff replied:
No, it's not deprecated and was introduced for the customizing part. I was wrong...it works for all possible ways:

```yaml
    json_attributes:
      - "time"
      - "date"
      - "milliseconds_since_epoch"
    json_attributes: ["time", "date", "milliseconds_since_epoch"]
    json_attributes: time, date, milliseconds_since_epoch
    json_attributes: time,date,milliseconds_since_epoch
    json_attributes: "time", "date", "milliseconds_since_epoch"
```
But I still would prefer if we don't promote `a, b, c` because it will lead to confusion for users which are not that familiar with YAML and are looking at other components/platforms.

nickovs replied:
Great. The documentation pull request that I submitted uses option `d` for the example but I'm happy to document it with whichever form you think should be promoted.

nickovs committed:
Added an explicit default value to the json_attributes config entry.

nickovs committed:
Removed self.update() from __init__() body.

nickovs committed:
Merge branch 'rest-json-attrs' of github.com:nickovs/home-assistant into rest-json-attrs

nickovs committed:
Expended unit tests for error cases of json_attributes processing.

fabaff committed:
Align quotes

MartinHjelmare commented:
Looks like this branch history was messed up. We should probably revert it.
